### PR TITLE
Payjoin: Better UIH1 & UIH2 based selection

### LINKED ISSUE #1470: Improve Payjoin utxo selector
Currently it only takes into consideration all outputs + the utxo you are considering.
Since we now have access to the values of each input of the original tx, we can expand this properly. 
Further, if we cannot find a single utxo to match the criteria, we can attempt to add multiple utxos.

### PR DESCRIPTION
**Pull request by Kukks**
closes #1470 

~I need to add some tests.~

### CODE REVIEW

File: BTCPayServer/Payments/PayJoin/PayJoinEndpointController.cs [hunk 1]
```diff
@@ -275,8 +275,9 @@ async Task UnlockUTXOs()
                 var prevOuts = originalTx.Inputs.Select(o => o.PrevOut).ToHashSet();
                 utxos = utxos.Where(u => !prevOuts.Contains(u.Outpoint)).ToArray();
                 Array.Sort(utxos, UTXODeterministicComparer.Instance);
-                foreach (var utxo in await SelectUTXO(network, utxos, output.Value,
-                    psbt.Outputs.Where(o => o.Index != output.Index).Select(o => o.Value).ToArray()))
+                
+                foreach (var utxo in (await SelectUTXO(network, utxos, psbt.Inputs.Select(input => input.WitnessUtxo.Value.ToDecimal(MoneyUnit.BTC)),  output.Value.ToDecimal(MoneyUnit.BTC),
+                    psbt.Outputs.Where(psbtOutput => !psbtOutput.Equals(output)).Select(psbtOutput => psbtOutput.Value.ToDecimal(MoneyUnit.BTC)))).selectedUTXO)
```
Comment (NicolasDorier) [inline BTCPayServer/Payments/PayJoin/PayJoinEndpointController.cs:280]: Why use `Equals` this is not well defined.

File: BTCPayServer/Payments/PayJoin/PayJoinEndpointController.cs [hunk 2]
```diff
@@ -504,42 +505,51 @@ private JObject CreatePayjoinError(int httpCode, string errorCode, string friend
...
                 }
 
+                if (invalid)
+                {
+                    continue;
+                }
                 if (await _payJoinRepository.TryLock(availableUtxo.Outpoint))
                 {
-                    return new UTXO[] { availableUtxo };
+                    return (new[] {availableUtxo}, false);
```
Comment (NicolasDorier) [inline BTCPayServer/Payments/PayJoin/PayJoinEndpointController.cs:550]: this is actually randomized, the result depends on how `availableUtxos` are sorted, and they are sorted randomly.
You should rename this if you use it in the tests. (Like an enum SelectedType with value like `UIH1,UIH2,Ordered`)

Comment (Kukks) [reply BTCPayServer/Payments/PayJoin/PayJoinEndpointController.cs:550]: yeah good point, this is specifically for tests

File: BTCPayServer/Payments/PayJoin/PayJoinEndpointController.cs [hunk 3]
```diff
@@ -275,8 +275,9 @@ async Task UnlockUTXOs()
                 var prevOuts = originalTx.Inputs.Select(o => o.PrevOut).ToHashSet();
                 utxos = utxos.Where(u => !prevOuts.Contains(u.Outpoint)).ToArray();
                 Array.Sort(utxos, UTXODeterministicComparer.Instance);
-                foreach (var utxo in await SelectUTXO(network, utxos, output.Value,
-                    psbt.Outputs.Where(o => o.Index != output.Index).Select(o => o.Value).ToArray()))
+                
+                foreach (var utxo in (await SelectUTXO(network, utxos, psbt.Inputs.Select(input => input.WitnessUtxo.Value.ToDecimal(MoneyUnit.BTC)),  output.Value.ToDecimal(MoneyUnit.BTC),
+                    psbt.Outputs.Where(psbtOutput => psbtOutput != output).Select(psbtOutput => psbtOutput.Value.ToDecimal(MoneyUnit.BTC)))).selectedUTXO)
```
Comment (NicolasDorier) [inline BTCPayServer/Payments/PayJoin/PayJoinEndpointController.cs:280]: why do you even change this?

Comment (Kukks) [reply BTCPayServer/Payments/PayJoin/PayJoinEndpointController.cs:280]: the params separate the payment output from the others to compute what the modified payment output value will be

Comment (NicolasDorier) [reply BTCPayServer/Payments/PayJoin/PayJoinEndpointController.cs:280]: I know but before I was comparing the index, why did you modified to compare the object?

Comment (Kukks) [reply BTCPayServer/Payments/PayJoin/PayJoinEndpointController.cs:280]: ah, no clue why. Does it make a difference?

Comment (Kukks) [reply BTCPayServer/Payments/PayJoin/PayJoinEndpointController.cs:280]: reverted back t yours

Comment (NicolasDorier) [reply BTCPayServer/Payments/PayJoin/PayJoinEndpointController.cs]: I don't think so, but I had a doubt that I override the `.Equals` (which was a bad idea), I prefer not touching what does not break.

**Pull request merged**